File: PVSS.py
from web3 import Web3

w3 = Web3(Web3.HTTPProvider('http://127.0.0.1:7545'))
from solcx import compile_standard, install_solc

install_solc("0.8.0")
import json  # to save the output in a JSON file
import random
import util
import secrets
import sympy  # consider removing this dependency, only needed for mod_inverse
from py_ecc.bn128 import G1, G2
from py_ecc.bn128 import add, multiply, neg, pairing, is_on_curve
from py_ecc.bn128 import curve_order as CURVE_ORDER
from py_ecc.bn128 import field_modulus as FIELD_MODULUS
from typing import Tuple, Dict, List, Iterable, Union

with open("contracts/DAOsForVote.sol", "r") as file:
    contact_list_file = file.read()

compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"DAOsForVote.sol": {"content": contact_list_file}},
        "settings": {
            "outputSelection": {
                "*": {
                    "*": ["abi", "metadata", "evm.bytecode", "evm.bytecode.sourceMap"]
                    # output needed to interact with and deploy contract
                }
            }
        },
    },
    solc_version="0.8.0",
)

# print(compiled_sol)
with open("compiled_code.json", "w") as file:
    json.dump(compiled_sol, file)
# get bytecode
bytecode = compiled_sol["contracts"]["DAOsForVote.sol"]["DAOsForVote"]["evm"]["bytecode"]["object"]
# get abi
abi = json.loads(compiled_sol["contracts"]["DAOsForVote.sol"]["DAOsForVote"]["metadata"])["output"]["abi"]
# Create the contract in Python
contract = w3.eth.contract(abi=abi, bytecode=bytecode)

chain_id = 5777
accounts0 = w3.eth.accounts[0]
transaction_hash = contract.constructor().transact({'from': accounts0})
# 等待合约部署完成
transaction_receipt = w3.eth.wait_for_transaction_receipt(transaction_hash)
# 获取部署后的合约地址
contract_address = transaction_receipt['contractAddress']
# print("合约已部署，地址：", contract_address)
Contract = w3.eth.contract(address=contract_address, abi=abi)

keccak_256 = Web3.solidityKeccak
"""
H0=(9727523064272218541460723335320998459488975639302513747055235660443850046724,5031696974169251245229961296941447383441169981934237515842977230762345915487)
H1=(5031696974169251245229961296941447383441169981934237515842977230762345915487,9727523064272218541460723335320998459488975639302513747055235660443850046724)
"""
H1 = multiply(G1, 9868996996480530350723936346388037348513707152826932716320380442065450531909)  # 生成元H1


def random_scalar() -> int:  # Generate random numbers
    """ Returns a random exponent for the BN128 curve, i.e. a random element from Zq.
    """
    return secrets.randbelow(CURVE_ORDER)


def Setup(n, t):  # PVSS Key Generation   #PVSS的公私钥产生，注意有效密钥从0开始索引
    """ Generates a random keypair on the BN128 curve.
        The public key is an element of the group G1.
        This key is used for deriving the encryption keys used to secure the shares.
        This is NOT a BLS key pair used for signing messages.
    """
    sk = []
    pk = []
    sk.extend([random_scalar() for i in range(0, n)])
    pk.extend([multiply(G1, sk[i]) for i in range(0, n)])  # only need 1-n

    return {"pk": pk, "sk": sk}


def share_secret(secret: int, sharenum: int, threshold: int) -> Dict[
    int, int]:  # PVSS.Share(s,n,t) n is sharenum value .t is threshold value
    coefficients = [secret] + [random_scalar() for j in range(threshold)]

    # Shamir 分享
    def f(x: int) -> int:
        """ evaluation function for secret polynomial
        """
        return (
                sum(coef * pow(x, j, CURVE_ORDER) for j, coef in enumerate(coefficients)) % CURVE_ORDER
        )

    indices = [i for i in range(1, sharenum + 1)]
    shares = {x: f(x) for x in indices}
    return shares


def Share(s_j, H1, pk, n, t):
    SSShare = share_secret(s_j, n, t)  # voter PVSS.share=(v,c)
    # 注意数组第一位为0，v，c数组长度为n+1
    v = [0]
    c = [0]
    DLEQ_Proof = [0]
    v.extend([multiply(H1, SSShare[i + 1]) for i in range(0, n)])  # v_i=H1^s_i
    c.extend([multiply(pk[i], SSShare[i + 1]) for i in range(0, n)])  # c_i=pk_i^s_i

    DLEQ_Proof.extend([DLEQ(H1, v[i + 1], pk[i], c[i + 1], SSShare[i + 1]) for i in range(0, n)])
    # DLEQ的Proof为证明v，c确实是由该多项式f(x)所生成,例如s_i=f(i)。
    res = {"v": v[1:], "c": c[1:], "DLEQ_Proof": DLEQ_Proof[1:]}
    return res


def DLEQ(x1, y1, x2, y2, alpha: int) -> Tuple[int, int]:  # 生成DLEQ承诺 alpha为需要承诺的零知识证明的值
    """ DLEQ... discrete logarithm equality
        Proofs that the caller knows alpha such that y1 = x1**alpha and y2 = x2**alpha
        without revealing alpha.
    """
    w = random_scalar()
    a1 = multiply(x1, w)
    a2 = multiply(x2, w)
    c = keccak_256(
        abi_types=["uint256"] * 12,
        values=[
            int(v)
            for v in (a1)
                     + (a2)
                     + (x1)
                     + (y1)
                     + (x2)
                     + (y2)
        ],
    )
    c = int.from_bytes(c, "big")
    r = (w - alpha * c) % CURVE_ORDER
    return c, r


def DLEQ_verify(x1, y1, x2, y2, challenge: int, response: int) -> bool:  # DLEQ_Verify的链下验证函数
    a1 = add(multiply(x1, response), multiply(y1, challenge))
    a2 = add(multiply(x2, response), multiply(y2, challenge))
    c = keccak_256(  # pylint: disable=E1120
        abi_types=["uint256"] * 12,  # 12,
        values=[
            int(v)
            for v in (a1)
                     + (a2)
                     + (x1)
                     + (y1)
                     + (x2)
                     + (y2)
        ],
    )
    c = int.from_bytes(c, "big")
    return c == challenge


# 删除PVSS.DVerify和PVSS.PVerify，将此部分转移到链上进行

def Decrypt(c_ji, sk_i):  # PVSS.Decrypt，解密函数
    sh1_ji = multiply(c_ji, sympy.mod_inverse((sk_i) % CURVE_ORDER, CURVE_ORDER))
    return sh1_ji


def Reconstruct(res, n, t):  # PVSS.Reconstruct  秘密恢复函数
    recIndex = [i + 1 for i in range(0, t + 1)]
    print(recIndex)
    sum = multiply(H1, 0)

    for i in recIndex:
        print("i", i, util.lagrange_coefficient(i))
        sum = add(sum, multiply(res["v"][i], util.lagrange_coefficient(i)))
    return sum


def ShareForCandidates(s_j, H1, pk, n, t):
    SSShare = share_secret(s_j, n, t)  # voter PVSS.share=(v,c)
    # 注意数组第一位为0，v，c数组长度为n+1
    v = [0]
    c = [0]
    DLEQ_Proof = [0]
    v.extend([multiply(H1, SSShare[i + 1]) for i in range(0, n)])  # v_i=H1^s_i
    c.extend([multiply(pk[i], SSShare[i + 1]) for i in range(0, n)])  # c_i=pk_i^s_i

    DLEQ_Proof.extend([DLEQ(H1, v[i + 1], pk[i], c[i + 1], SSShare[i + 1]) for i in range(0, n)])
    # DLEQ的Proof为证明v，c确实是由该多项式f(x)所生成,例如s_i=f(i)。
    res = {"v": v[1:], "c": c[1:], "DLEQ_Proof": DLEQ_Proof[1:], "P_j": SSShare}
    return res

    # reed solomon check


def RScodeVerify(res):
    # print(len(res))
    def coefficient(i: int) -> int:
        result = 1
        for j in range(1, len(res) + 1):
            # print(j)
            # j=j-1
            if i != j:
                # print(i - j) #% CURVE_ORDER)
                result *= sympy.mod_inverse((i - j) % CURVE_ORDER, CURVE_ORDER)
                result %= CURVE_ORDER
        return result

    codeword = []
    for i in range(1, len(res) + 1):
        codeword.append(coefficient(i) % CURVE_ORDER)
    sum = multiply(H1, 1)
    for i in range(0, len(res)):
        sum = add(sum, multiply(res[i], codeword[i]))
    if (sum == H1):
        return 1
    else:
        return 0
    #   sum=add(sum, multiply(res["v"][i], lagrange_coefficient(i)))


# 留作测试所用
"""
if __name__ == '__main__':
    key = Setup(10, 5)

    n = 10
    t = 5
    print("answer")
    print(multiply(H1, 233333))
    print(".........")

    res1 = Share(233333, H1, key["pk"], n, t)
    #print(res1)
    #print(RScodeVerify(res1["v"]))
    print(Contract.functions.inv(0,CURVE_ORDER).call())
    print(sympy.mod_inverse((1) % CURVE_ORDER, CURVE_ORDER))
    #sum = Reconstruct(res1, 10, 5)
    #print(sum)
    # Reconstruct2(res1,10,5)
    def test():
        for i in range((CURVE_ORDER-30),(CURVE_ORDER+31)):

            print(i,"..........",sympy.mod_inverse((i) % CURVE_ORDER, CURVE_ORDER))

    #test()
"""

File: go/crypto/PVSS/PVSS.go
package PVSS

import (
	"crypto/rand"
	"crypto/sha256"

	//"fmt"
	"bytes"
	"math/big"

	bn256 "github.com/ethereum/go-ethereum/crypto/bn256/google"
)

type DLEQ struct {
	C  *big.Int
	Z  *big.Int
	RG *bn256.G1
	RH *bn256.G1
}

// 需要运算的参数
type SecretSharing struct {
	//Fi []*big.Int
	V      []*bn256.G1
	C      []*bn256.G1
	Proofs []DLEQ
}

// PVSS.Setup
func Setup(numTalliers int) ([]*big.Int, []*bn256.G1) {
	sks := make([]*big.Int, numTalliers)
	pks := make([]*bn256.G1, numTalliers)
	for i := 0; i < numTalliers; i++ {
		sks[i], _ = rand.Int(rand.Reader, bn256.Order)
		pks[i] = new(bn256.G1).ScalarBaseMult(sks[i])
	}
	return sks, pks
}

func DLEQProof(G, H *bn256.G1, xG, xH *bn256.G1, x *big.Int) DLEQ {
	//生成承诺
	r, _ := rand.Int(rand.Reader, bn256.Order)
	rG := new(bn256.G1).ScalarMult(G, r)
	rH := new(bn256.G1).ScalarMult(H, r)

	// 计算挑战
	new_hash := sha256.New()
	new_hash.Write(xG.Marshal())
	new_hash.Write(xH.Marshal())
	new_hash.Write(rG.Marshal())
	new_hash.Write(rH.Marshal())

	cb := new_hash.Sum(nil)
	c := new(big.Int).SetBytes(cb)
	c.Mod(c, bn256.Order)

	// 生成相应
	z := new(big.Int).Mul(c, x)
	z.Sub(r, z)
	z.Mod(z, bn256.Order)

	return DLEQ{
		C:  c,
		Z:  z,
		RG: rG,
		RH: rH,
	}
}

// PVSS.Share: Generate PVSS Shares({Vj},{Cj},DLEQ proofs)
func Share(secret *big.Int, h *bn256.G1, pks []*bn256.G1, threshold, numShares int) *SecretSharing {
	coefficients := make([]*big.Int, threshold)
	coefficients[0] = secret
	for i := 1; i < threshold; i++ {
		coefficients[i], _ = rand.Int(rand.Reader, bn256.Order)
	}
	shares := make([]*big.Int, numShares)
	for i := 0; i < numShares; i++ {
		x := big.NewInt(int64(i + 1))
		shares[i] = EvaluatePolynomial(coefficients, x, bn256.Order)
	}

	v := make([]*bn256.G1, numShares)
	for i := 0; i < numShares; i++ {
		v[i] = new(bn256.G1).ScalarMult(h, shares[i])
	}

	//Generate the encrypted shares under talliers' public key
	c := make([]*bn256.G1, numShares)
	for i := 0; i < numShares; i++ {
		c[i] = new(bn256.G1).ScalarMult(pks[i], shares[i])
	}

	//Generate the DLEQ proofs for each encrypted shares
	proofs := make([]DLEQ, numShares)
	for i := 0; i < numShares; i++ {
		proofs[i] = DLEQProof(h, pks[i], v[i], c[i], shares[i])
	}

	return &SecretSharing{
		//Shares: shares,
		V:      v,
		C:      c,
		Proofs: proofs,
	}
}

// Verify verifies the DLEQ proof
func DLEQVerify(c, z *big.Int, G, H, xG, xH, rG, rH *bn256.G1) bool {
	zG := new(bn256.G1).ScalarMult(G, z)
	zH := new(bn256.G1).ScalarMult(H, z)
	cxG := new(bn256.G1).ScalarMult(xG, c)
	cxH := new(bn256.G1).ScalarMult(xH, c)
	a := new(bn256.G1).Add(zG, cxG)
	b := new(bn256.G1).Add(zH, cxH)
	if !(rG.String() == a.String() && rH.String() == b.String()) {
		return false
	}
	return true
}

// Reed Solomon check
func coefficient(i int, n int) *big.Int {
	result := big.NewInt(1)
	for j := 1; j < n+1; j++ {
		if i != j {
			result = result.Mul(result, new(big.Int).ModInverse(new(big.Int).Sub(big.NewInt(int64(i)), big.NewInt(int64(j))), bn256.Order))
			result = result.Mod(result, bn256.Order)
		}
	}
	return result
}

func RScodeVerify(shares []*bn256.G1, H1 *bn256.G1) bool {
	codeword := make([]*big.Int, len(shares))
	for i := 1; i <= len(shares); i++ {
		codeword[i-1] = new(big.Int).Mod(coefficient(i, len(shares)), bn256.Order)
	}
	sum := new(bn256.G1).ScalarMult(H1, big.NewInt(1))
	for i := 0; i < len(shares); i++ {
		sum = new(bn256.G1).Add(sum, new(bn256.G1).ScalarMult(shares[i], codeword[i]))
	}
	if bytes.Equal(sum.Marshal(), H1.Marshal()) {
		return true // if verify pass return true
	}
	return false // else return false
}

// PVSS.DVerify
func DVerify(secretsharing *SecretSharing, h *bn256.G1, pks []*bn256.G1) bool {
	for i := 0; i < len(secretsharing.V); i++ {
		if !DLEQVerify(secretsharing.Proofs[i].C, secretsharing.Proofs[i].Z, h, pks[i], secretsharing.V[i], secretsharing.C[i], secretsharing.Proofs[i].RG, secretsharing.Proofs[i].RH) {
			return false
		}
	}
	if !RScodeVerify(secretsharing.V, h) {
		return false
	}
	return true
}

// PVSS.Decrypt
func Decrypt(h *bn256.G1, pk *bn256.G1, c *bn256.G1, sk *big.Int) (*bn256.G1, DLEQ) {
	sh := new(bn256.G1).ScalarMult(c, sk)
	proof := DLEQProof(h, pk, sh, c, sk)
	return sh, proof
}

// PVSS.PVerify
func PVerify(h *bn256.G1, pk *bn256.G1, c *bn256.G1, sh *bn256.G1, proof *DLEQ) bool {
	if !DLEQVerify(proof.C, proof.Z, h, pk, sh, c, proof.RG, proof.RH) {
		return false
	}
	return true
}

// evaluatePolynomial 在给定的 x 处计算多项式的值
func EvaluatePolynomial(coefficients []*big.Int, x, order *big.Int) *big.Int {
	result := new(big.Int).Set(coefficients[0])
	xPower := new(big.Int).Set(x)

	for i := 1; i < len(coefficients); i++ {
		term := new(big.Int).Mul(coefficients[i], xPower)
		term.Mod(term, order)
		result.Add(result, term)
		result.Mod(result, order)
		xPower.Mul(xPower, x)
		xPower.Mod(xPower, order)
	}

	return result
}

// lagrangeInterpolation 求拉格朗日插值法的系数
func LagrangeCoefficient(l *big.Int, indices []*big.Int, threshold int) []*big.Int {
	// k是分享的数量
	coefficient := make([]*big.Int, threshold)

	// 对于每个分享
	for i := 0; i < threshold; i++ {
		// 初始化分子（num）和分母（den）为1
		num := big.NewInt(1)
		den := big.NewInt(1)

		// 计算拉格朗日基函数的分子和分母
		for j := 0; j < threshold; j++ {
			if i != j {
				// 分子累乘 -indices[j]
				num.Mul(num, new(big.Int).Sub(l, indices[j]))
				num.Mod(num, bn256.Order)

				// 分母累乘 indices[i] - indices[j]
				den.Mul(den, new(big.Int).Sub(indices[i], indices[j]))
				den.Mod(den, bn256.Order)
			}
		}

		// 计算分母的逆元（模order）
		den.ModInverse(den, bn256.Order)
		// 计算每一项的值 shares[i] * num * den
		term := new(big.Int).Mul(num, den)
		term.Mod(term, bn256.Order)
		coefficient[i] = term
	}
	return coefficient
}

func Reconstruct(coefficient []*big.Int, sh []*bn256.G1) *bn256.G1 {
	secret := new(bn256.G1).ScalarBaseMult(big.NewInt(0))
	for i := 0; i < len(coefficient); i++ {
		secret = new(bn256.G1).Add(secret, new(bn256.G1).ScalarMult(sh[i], coefficient[i]))
	}
	return secret
}
